memory_w25q1128jv_spi/memory_w25q1128jv_spi.cc

```diff
@@ -0,0 +1,220 @@
+            _state = State::PAGE_WRITE;
+
+            /* Set Write Enable to 1 */
+            uint8_t CHIP_ENABLE[1] = {0x06};
+
+            // chip enable instruction to enable chip
+            HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+            if (HAL_SPI_Transmit(_hspi, CHIP_ENABLE, 1, SPI_Timeout) != HAL_OK) {
+                _state = State::ERROR;
+                break;
+            }
+            HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+        }
+
+        case State::PAGE_WRITE: {
+            // Checks if write enable latch is = 1
+            // write enable latch is 1st bit in status reg
+            if ((MemoryW25q1128jvSpi::ReadStatusReg1() & 0x02) != 2) break;
+
+            _state = State::COMPLETE;
+
+            /* Write the 64 bytes into the memory  */
+            uint8_t PAGE_PROGRAM[4] = {0x02, (uint8_t)(address >> 16), (uint8_t)(address >> 8), (uint8_t)(address)};
+
+            // Write 64 byte Data  into memory
+            HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+            if (HAL_SPI_Transmit(_hspi, PAGE_PROGRAM, 4, SPI_Timeout) != HAL_OK) {
+                _state = State::ERROR;
+                break;
+            }
+            if (HAL_SPI_Transmit(_hspi, data, 64, SPI_Timeout) != HAL_OK) {
+                _state = State::ERROR;
+                break;
+            }
+            HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+        }
+
+        case State::COMPLETE: {
+            // Checks if busy bit is = 0
+            // write busy is 1st bit in status reg
+            if ((MemoryW25q1128jvSpi::ReadStatusReg1() & 0x01) != 0) break;
+
+            _state = State::CHIP_ENABLE;
+
+            // increments to the next 64 byte address
+            address += 64;
+
+            break;
+        }
+
+        default:
+            break;
+    }
+    return _state;
+}
+
+int MemoryW25q1128jvSpi::ChipRead(uint32_t readAddress, uint8_t (&chipData)[64]) {
+    // Control pins his for no false memory actions
+    HAL_GPIO_WritePin(_wpPort, _wpPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_holdPort, _holdPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // initalisation of input, output, and dumby telemtrydata variable for temporary writing to DataBundle
+    uint8_t READ_DATA[4] = {0x03, (uint8_t)(readAddress >> 16), (uint8_t)(readAddress >> 8), (uint8_t)(readAddress)};
+
+    // Reads address page in memory
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, READ_DATA, 4, SPI_Timeout) != HAL_OK) return -1;
+    if (HAL_SPI_Receive(_hspi, chipData, 64, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // Waits until busy bit is = 0
+    // write busy is 1st bit in status reg
+    while ((MemoryW25q1128jvSpi::ReadStatusReg1() & 0x01) != 0) {
+    }
+
+    return 0;
+}
+
+int MemoryW25q1128jvSpi::ChipReadDump(uint8_t (&chipData)[64]) {
```

Comment:
im confused by this method, it always just reads block 0?

Reply:
`static uint32_t dumpAddress = 0;` basically acts like a private variable in this function. So, its value is initialised as 0 once and then in the next call to this function, the variable's value from the last function call is remembered while the variable stays in the scope of the function. I can assure you that it does increment properly for every function call.